Step3 리뷰 요청 드립니다.

itdar opened the pull request:
안녕하세요!~
2단계 이후로 바빠서 좀 늦었네요 ㅎㅎ;
점수 계산은 각각 프레임이 자신들의 점수 계산을 할 수 있게끔 해두었고,
실제 누적되면서 보이는 점수는 ui 단에서 합산해서 보여지도록 만들어두었습니다.
리뷰 감사합니다!~

itdar committed:
refactor: 선동님 2단계 리뷰사항 수정 커밋

- 파라미터테스트로 테스트 정리
- 거터 심볼 추가
- 의미에 안맞는 테스트 fix
- ui 출력에 뉴라인 추가

itdar committed:
feat: 볼링점수 구현 중 일반계산인지, 추가점수 있는지 확인하는 유틸클래스 구현

itdar committed:
chore: 볼링 점수 다시 구현 전 이전 기본기능 동작하도록 수정

itdar committed:
feat: 각 프레임별 점수 계산 기능 구현 완료 (일반 / 최종 프레임)

itdar committed:
feat: ui 수정 중 커밋

itdar committed:
fix: 마지막프레임 계산 시 일반프레임 일부 출력 안되는 버그 수정

itdar committed:
refactor: 전체적인 리팩토링

- 인덴테이션레벨 1로 맞춤
- 테스트 수정
- 사용하지 않는 코드 제거
- 불필요한 주석 제거
등

seondongpyo commented on src/main/java/bowling/domain/BonusShotCount.java:
```diff
@@ -0,0 +1,29 @@
+    }
+
+    public void minus() {
+        if (count > 0)
+            --count;
```
자바 코드 컨벤션에 따라, `if`문 작성 시 괄호를 추가해주면 좋을 것 같아요.
https://google.github.io/styleguide/javaguide.html#s4.1.1-braces-always-used

itdar replied:
옙 링크 감사합니다
추가해보겠습니다!!

seondongpyo commented on src/main/java/bowling/domain/HitState.java:
```diff
@@ -1,24 +1,53 @@
+        }
+        return Optional.of(next.firstScore() + 10);
+    }),
+    STRIKE(frame -> {
+        Frame current = frame.next();
+        if (current == null || current.isNotThrowFirstYet()) {
+            return Optional.empty();
+        }
+
+        TotalScore totalScore = TotalScore.ofInitialStrike();
+        BonusShotCount bonusCount = BonusShotCount.ofStrike();
+        while (current != null && bonusCount.isRemained()) {
+            totalScore.appendScoreOf(bonusCount, current);
+            current = current.next();
+        }
+
+        return resultScoreAsOptional(totalScore, bonusCount);
+    });
```
`BonusShotCount`, `TotalScore`는 스트라이크를 친 경우에만 한하여 보너스 점수를 계산하기 위해 
추가된 것으로 보이는데, 개인적으로는 점수 계산을 위해 너무 많은 클래스가 관여하는 게 아닌가 싶어요.
다음 3단계 요구사항의 힌트를 참고하여, 점수를 `Frame`에서 구해보도록 하는 건 어떨까요?
상태 패턴을 참고해보시는 것도 좋을 것 같습니다.
```
- 점수를 구하는 역할을 각 Frame이 담당할 수 있도록 구현해 본다. 
- Frame 외부에서 점수를 계산해 set하는 것이 아니라 Frame 자체가 점수를 계산할 수 있도록 한다.
```

itdar replied:
옙 인덴테이션 레벨을 낮추려다보니 좀 억지로 만드는 느낌이 들긴 했었는데요 ㅎㅎ;;
프레임 내부로 넣어보겠습니다!!
감사합니다

itdar replied:
수정하려고 기존 코드 작성 시를 다시 생각하다보니
Frame 내부에서 계산을 하도록 구현했었습니다 ㅎㅎ; 
Frame 의 scoreCalculated() 를 통해서 해당 프레임의 점수를 계산해서 가지고 오도록 하는 메서드를 만들었던 것이었습니다.

말씀해주신 상태패턴처럼 frame.hit() 로 상태가 확정되는 상황에서 인스턴스 변수로 HitState (상태) 를 가지고 있다가 계산시에 사용할까 했었으나.. 상태를 확인하는 코드는 어디에서든 들어가야하고, 인스턴스변수를 굳이 추가해야하나 해서.. Frame 으로 점수계산 메시지를 보냈을 때 enum 에 위임된 계산로직을 사용하는 현재 상태로 만들게 되었습니다.
클래스가 추가된 것은 점수계산에만 쓰이는 클래스인데, Frame 이나 enum 클래스 안으로 로직을 빼기보다는 클래스를 private inner 로 변경해보았습니다. 이렇게는 어떻게 생각하시나요? @seondongpyo

seondongpyo replied:
질문에 대한 지극히 개인적인 의견을 드리니 참고만 부탁드리겠습니다.
점수 계산을 위한 객체들에게 `Frame` 객체가 전달되고, 이후 `Frame`으로부터 점수를 꺼내서 계산하고 있는데,
대신에 실제 점수를 가지고 있는 `Score`에게 적절한 역할과 책임을 부여하는 방향으로 진행해보시면 어떨까 생각해요.
이 부분은 다음 단계 진행하시면서 리팩토링 힌트를 한 번 참고해보시는 것도 좋을 것 같습니다.

itdar replied:
옙 다음 단계 진행전에 힌트 읽어보고 적용해보겠습니다!
감사합니다!~

itdar replied:
옙 다음 단계 진행전에 힌트 읽어보고 적용해보겠습니다!
감사합니다!~ @seondongpyo

seondongpyo commented on src/main/java/bowling/domain/Score.java:
```diff
@@ -36,4 +38,11 @@ public int hashCode() {
         return Objects.hash(score);
     }
 
+    public boolean isHitAll() {
+        return score == START_PIN_COUNT;
+    }
```
사용하지 않는 메서드는 제거해보면 어떨까요?

itdar replied:
엇 ㅎㅎ; 지워두겠습니다!~

seondongpyo commented on src/main/java/bowling/domain/Score.java:
```diff
@@ -36,4 +38,11 @@ public int hashCode() {
+        return score == START_PIN_COUNT;
+    }
+
+    public boolean isHitBy(int hitScore) {
+        return this.score == hitScore;
+    }
```
테스트에서만 사용되는 코드를 프로덕션 코드에 추가하는 건 지양해야 한다고 생각해요.
프로덕션 코드가 테스트 코드에 의존적이라고 볼 수 있기 때문인데, 어떻게 생각하시나요?

itdar replied:
음 저도 테스트만을 위한 코드는 최대한 배제하는 것에 동의합니다!! 다만 접근레벨에 대해서는 protected 정도로 열어두는건 테스트에 필요하다면 큰 거부감없이 테스트에 사용하는 편입니다!
이 메서드 같은 경우에는 tdd 로 진행하면서 클래스 만들었을 때 처음에 만들어두었다가 사용을 안하게 된 코드네요 ㅎㅎ;;
제거해보겠습니다!~

seondongpyo commented on src/main/java/bowling/ui/InputCui.java:
```diff
@@ -16,7 +16,7 @@ public static String inputParticipants() {
 
         if (!isValidNames(nameInitial)) {
-            out.printf("플레이어의 이름 길이는 %d 의 문자열만 가능합니다.", NAME_LENGTH);
+            out.printf("플레이어의 이름 길이는 %d 의 문자열만 가능합니다.\n", NAME_LENGTH);
```
플랫폼에 독립적인 줄바꿈 문자를 사용해보는 건 어떨까요?
https://www.baeldung.com/java-string-newline#3-using-platform-independent-newline-characters

itdar replied:
옙 수정해보겠습니다!!

seondongpyo commented on src/main/java/bowling/ui/OutputCui.java:
```diff
@@ -22,84 +24,126 @@ public static void drawStatus(Game game) {
+        return frame.scoreCalculated();
+    }
+
+    private static String drawFirstScore(Optional<Score> firstScore) {
```
`Optional`을 반환 타입이 아닌 다른 목적으로 사용할 경우의 문제점을 참고해보시면 좋을 것 같아요.
https://catsbi.oopy.io/81ee5bdc-6825-46f8-b1d2-c608ab2d6465

itdar replied:
엇 그러네요 ㅎㅎ;; 수정해보겠습니다 감사합니다!!

seondongpyo commented on src/test/java/bowling/domain/FrameTest.java:
```diff
@@ -80,4 +80,38 @@ void isDoneTest2() {
+    @DisplayName("스트라이크 추가 점수는 해당 프레임이 끝나고 다음 투구 2개의 점수가 계산된 후 계산된다.")
+    @Test
+    void calculateScoreTest3() {
+        frame.shot(9);
+        frame.shot(0);
+        Frame second = frame.createNext();
```
스트라이크를 테스트하기 때문에 해당 코드는 없어도 되지 않을까요?

itdar replied:
엇 곧바로 스트라이크 프레임으로 시작하도록 수정해보겠습니다!!

seondongpyo commented on src/main/java/bowling/ui/OutputCui.java:
```diff
@@ -22,84 +24,126 @@ public static void drawStatus(Game game) {
         out.print(GAP + SCORE_SEPARATOR);
     }
 
     private static void drawFinalFrame(Frame finalFrame) {
```
마지막 프레임에서 두 번째와 세 번째 투구 결과 출력을 확인해주시면 좋을 것 같아요.

![image](https://user-images.githubusercontent.com/64854054/170847267-5a998c4b-aca6-4f5d-a8f1-ecbcfb4339dc.png)

itdar replied:
아 밀리는게 있네요 ㅎㅎ 수정해보겠습니다!~

seondongpyo reviewed (changes requested):
3단계 미션도 잘 작성해주셨네요 👍
몇 가지 코멘트 드렸는데 확인해주시면 좋을 것 같아요.
조금만 더 힘내시고 남은 주말도 잘 보내시길 바랍니다 🔥

itdar committed:
refactor: 3단계 선동님 리뷰사항 적용 리팩토링

- 프레임 점수계산에 사용되는 클래스를 계산로직이 위임되는 HitState enum 내부클래스로 이동
- Optional 을 파라미터에 사용한 것 제거
- 사용하지 않는 메서드 제거
- 코드 컨벤션에 안맞는 것 수정
- ouput ui 마지막 프레임 줄밀림 수정

seondongpyo commented on src/test/java/bowling/domain/FinalFrameTest.java:
```diff
@@ -0,0 +1,21 @@
+class FinalFrameTest {
+
+    @Test
+    void 마지막_프레임은_보너스점수_없이_다_던지면_점수가_바로_계산된다() {
```
마지막 프레임에서 스트라이크 또는 스페어를 치지 못한 경우도 테스트해보는 건 어떨까요?
보너스 투구를 얻지 못한 상황에서 한 번 더 투구를 하려고 한다면 예외가 발생할 것으로 보여요.
```
finalFrame.shot(5); // 첫 번째 투구
finalFrame.shot(4); // 두 번째 투구
finalFrame.shot(3); // 세 번째 투구?
```

itdar replied:
엇 그러네요 한번 추가해보겠습니다!!

seondongpyo reviewed (approved):
피드백 잘 반영해주셨습니다 👍
추가적인 코멘트와 함께 이번 단계는 여기서 merge 하겠습니다.
마지막 단계도 힘내시길 바랄게요 🔥
